sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -851,7 +851,7 @@ object PushDownPredicate extends Rule[LogicalPlan] with PredicateHelper {
 
     case filter @ Filter(condition, union: Union) =>
       // Union could change the rows, so non-deterministic predicate can't be pushed down
-      val (pushDown, stayUp) = splitConjunctivePredicates(condition).span(_.deterministic)
+      val (pushDown, stayUp) = splitConjunctivePredicates(condition).partition(_.deterministic)
```

Comment:
What does it mean "after the first non-deterministic"? Doesn't this simply partition predicates to deterministic and non-deterministic? Have it considered "first" non-deterministic?

Reply:
@viirya 
IIUC, with `span`, the deterministic predicates after the first non-deterministic will not be pushed down, but with `partition`, all deterministic predicates will be pushed down.